### DISCUSSION TITLE
Connecting via pgBouncer

### DISCUSSION THREAD

**Discussion by sauntimo** [supabase/supabase#8809] [Questions]
I'm trying to work out if I've set up pgBouncer correctly on supabase. I'm running an API on Google Cloud run, I have several container instances. I'm attempting to connect to supabase via pgBouncer, from prisma. The "real" problem I'm trying to solve is that I'm running out of connections in my prod env - which I was hoping pgBouncer would fix, so I feel like I am probably not using it correctly...

- Connection pooling is enabled
![Screenshot 2022-09-06 at 12 28 13](https://user-images.githubusercontent.com/2720466/188624627-4e107955-80c1-4e17-945f-befba7316795.png)

- connection string includes `:6543/postgres?pgbouncer=true`

- but looking at project/.../database/roles, I can see 17 connections for my "postgres" user, and 0 for pgBouncer -
![Screenshot 2022-09-06 at 12 29 11](https://user-images.githubusercontent.com/2720466/188624815-c5414aa6-d147-4834-b108-5dedae6279e7.png)
 
Is this what this is supposed to look like? To me it looks like these users are connecting directly to the database, rather than to pgBouncer.

- If I query `pg_stat_activity`, I can see connections for `usename` "postgres" which is used by my API contianers, and `client_addr` values representing a couple of different container instances. Are these "forwarded on" from pgBouncer? or have they by passed pgBouncer entirely? I am not familiar with what the either of these should look like if *were* using pgBouncer correctly so it's hard for me to tell what's going on. 

- Is there a way I can log in to pgBouncer running on supabase directly, and run commands like `SHOW POOLS;` so that I can try and work out if I am actually connection through pgBouncer?

Very grateful for any assistance.

**Comment by burmecia** [comment]
For pgBouncer connection pooling, some of the points I can think of:

- For connection through Prisma in serverless container, the recommended connection string is like `:6543/postgres?pgbouncer=true&connection_limit=1`. See more details: https://supabase.com/docs/guides/integrations/prisma#connection-pooling-with-supabase

- If you're connecting using pgBouncer, you should see some connections from it in https://app.supabase.com/project/_/database/roles, 
  <img width="745" alt="image" src="https://user-images.githubusercontent.com/19306324/188766467-cc873b49-68ec-41b7-b785-9d8576a3c593.png">

- Also in `pg_stat_activity`, you should see the `client_addr` is `127.0.0.1` which is from local pgBouncer. Client's real ip address won't be forwarded by pgBouncer.

- Unfortunately, for security reason, you cannot log in using pgBouncer user at this moment, so operations like `SHOW POOLS;` won't be available. 

BTW, could you double check if clients are connecting to correct db? I am suspecting they are still using direct db connection rather than through pgBouncer.

Hope this can help you.

**Reply by sauntimo** [reply]
Yeah this does help, thank you - my hypothesis at the moment is that the database URL from the build stage, which is direct to postgres and bypasses pgBouncer (so that prisma can run migrations) is still in the env when the build is complete and the URL which is injected from the container env vars is ignored. Thanks for clarifying the above, it's useful to know my suspicions were correct!
